File: day_03/day_03.py
import re

res1 = 0
res2 = 0
enabled = True

def is_valid_mul(index, line):
    pattern = r'^mul\((\d+),(\d+)\)'
    match = re.search(pattern, line[index:])
    if match:
        return int(match.group(1)) * int(match.group(2))
    return 0

def check_enabled(index, line):
    old_enabled = enabled
    if re.search(r'^do\(\)', line[index:]):
        return True
    if re.search(r'^don\'t\(\)', line[index:]):
        return False
    return old_enabled

with open('input.txt', 'r') as file:
    for line in file:
        for index, char in enumerate(line):
            res1 += is_valid_mul(index, line)
            enabled = check_enabled(index, line)
            if enabled:
                res2 += is_valid_mul(index, line)

print("#1 ->", res1)
print("#2 ->", res2)

File: day_03/day_03.cpp
#include <fstream>
#include <string>
#include <sstream>
#include <iostream>
#include <regex>

int MAX_LENGHT = 42;
bool enabled = true;
std::regex pattern_mul(R"(^mul\((\d+),(\d+)\))");
std::regex pattern_do(R"(^do\(\))");
std::regex pattern_dont(R"(^don\'t\(\))");

std::string readFile(const std::string& filename) {
	std::ifstream file(filename);
	std::stringstream buffer;
	buffer << file.rdbuf();
	return buffer.str();
}

int isMul(const std::string& input, size_t pos) {
	std::smatch match;

	if (std::regex_search(input.begin() + pos, input.begin() + pos + MAX_LENGHT, match, pattern_mul))
		return std::stoi(match[1]) * std::stoi(match[2]);
	return 0;
}


bool check_enabled(const std::string& input, size_t pos) {
	std::smatch match_do;
	std::smatch match_dont;

	if (std::regex_search(input.begin() + pos, input.begin() + pos + MAX_LENGHT, match_do, pattern_do))
		return true;
	if (std::regex_search(input.begin() + pos, input.begin() + pos + MAX_LENGHT, match_dont, pattern_dont))
		return false;
	return enabled;
}

int main() {
	int res1 = 0;
	int res2 = 0;
	std::string input = readFile("input.txt");

	for (int i = 0; i < input.length(); i++) {
		res1 += isMul(input, i);
		enabled = check_enabled(input, i);
		if (enabled)
			res2 += isMul(input, i);
	}

	std::cout << "#1 -> " << res1 << std::endl;
	std::cout << "#2 -> " << res2 << std::endl;
}